cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  using difference_type = ptrdiff_t;
+
+  using __env_t          = ::cuda::experimental::env_t<_Properties...>;
+  using __policy_t       = ::cuda::experimental::execution::execution_policy;
+  using __buffer_t       = ::cuda::experimental::uninitialized_async_buffer<_Tp, _Properties...>;
+  using __resource_t     = ::cuda::experimental::any_async_resource<_Properties...>;
+  using __resource_ref_t = _CUDA_VMR::async_resource_ref<_Properties...>;
+
+  // TODO: use an accessor that has host / device guardrails
+  using __mdspan_t = _CUDA_VSTD::mdspan<_Tp, _Extents, _LayoutPolicy>;
+
+  template <class, class, class, class...>
+  friend class async_mdarray;
+
+  // For now we require trivially copyable type to simplify the implementation
+  static_assert(_CCCL_TRAIT(_CUDA_VSTD::is_trivially_copyable, _Tp),
+                "cuda::experimental::async_mdarray requires T to be trivially copyable.");
+
+  static_assert(_CUDA_VSTD::__detail::__is_extents_v<_Extents>,
+                "mdspan's Extents template parameter must be a specialization of _CUDA_VSTD::extents.");
+
+  // At least one of the properties must signal an execution space
+  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
+                "The properties of cuda::experimental::async_mdarray must contain at least one execution space "
+                "property!");
+
+  //! @brief Convenience shortcut to detect the execution space of the async_mdarray
+  static constexpr bool __is_host_only = __select_execution_space<_Properties...> == _ExecutionSpace::__host;
+
+private:
+  __buffer_t __buf_{};
+  mapping_type __mapping_{};
+  __policy_t __policy_ = __policy_t::invalid_execution_policy;
+
+  //! @brief Helper to check container is compatible with this async_mdarray
+  template <class _Range>
+  static constexpr bool __compatible_range = _CUDA_VRANGES::__container_compatible_range<_Range, _Tp>;
+
+  //! @brief Helper to check whether a different async_mdarray still statisfies all properties of this one
+  template <class... _OtherProperties>
+  static constexpr bool __properties_match =
+    !_CCCL_TRAIT(_CUDA_VSTD::is_same,
+                 _CUDA_VSTD::__make_type_set<_Properties...>,
+                 _CUDA_VSTD::__make_type_set<_OtherProperties...>)
+    && _CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__make_type_set<_OtherProperties...>, _Properties...>;
+
+  //! @brief Helper to determine what cudaMemcpyKind we need to copy data from another async_mdarray with different
+  template <class... _OtherProperties>
+  static constexpr cudaMemcpyKind __transfer_kind =
+    __select_execution_space<_OtherProperties...> == _ExecutionSpace::__host
+      ? (__is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice)
+      : (__is_host_only ? cudaMemcpyDeviceToHost : cudaMemcpyDeviceToDevice);
+
+  //! @brief Helper to return an async_resource_ref to the currently used resource. Used to grow the async_mdarray
```

Comment:
Regarding "[u]sed to grow the `async_mdarray`," are you planning to make `async_mdarray` resizable after construction?

Reply:
This is a remnant from copying the code from vector.

Currently if we want to be able to assign one mdarray to another we need the ability to reallocate storage because they might have different sizes / memory resources

Reply:
So I would say we do not want to make it resizable but want to be able t assign